### PR TITLE
Use a different error type for operations on a local node

### PR DESCRIPTION
**Pull request by ma2bd**
## Motivation

The client code in `linera-core` is sometimes hard to debug because we're querying both validators and a local node with a similar API. In particular, the errors are displayed in the same way.

## Proposal

* Split `node.rs` and create `LocalNodeClient` in a new file `local_node.rs`
* Do not make `LocalNodeClient` implement the trait `ValidatorNode`
* Use a new error type `LocalNodeError` instead of `NodeError` in the APIs of `LocalNodeClient`
* Adjust the error messages so that it's clear when an error happened in a local node.

Next, this should allow us to finally use a proper error type instead of `anyhow` in `linera-core/src/client.rs` (https://github.com/linera-io/linera-protocol/issues/66)

## Test Plan

CI

## Release Plan

- [x] All good!
- [ ] Need to bump the major/minor version number in the next release of the crates.
- [ ] Need to update the developer manual.
- [ ] This PR is adding or removing Cargo features.
- [ ] Release is blocked and/or tracked by other issues (see links above)

## Reviewer Checklist

- [ ] The title and the summary of the PR are short and descriptive.
- [ ] The proposed solution achieves the goals stated in the PR.
- [ ] The test plan is reproducible and provides sufficient coverage.
- [ ] The release plan is adequate.
- [ ] The commits correspond to distinct logical changes.
- [ ] The code follows the [coding guidelines](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md).
- [ ] The proposed changes look correct.
- [ ] The CI is passing.
- [ ] All of the above!

### CODE REVIEW

File: linera-core/src/local_node.rs [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+                chain_id,
+                target_next_block_height,
+            })
+        }
+    }
+
+    /// Downloads and stores the specified blobs, unless they are already in the cache or storage.
+    ///
+    /// Does not fail if a blob can't be downloaded; it just gets omitted from the result.
+    pub async fn read_or_download_blobs<A>(
```
Comment (christos-h) [inline linera-core/src/local_node.rs:309]: From the name of the function I would not have guessed that if a blob failed to download it's just going to move on.

In which case maybe we want to change this to either:
1. The name of the function: `optimistically_read_or_download_blobs`
2. Return `Result<(Vec<HashedValue>, Vec<BytecodeLocation>)>` where the second entry in the tuple is the list of failed bytecodes.

Or both?

File: linera-core/src/local_node.rs [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+            }
+        }
+        drop(node); // Free the lock while awaiting the tasks.
+        if tasks.is_empty() {
+            return Ok(blobs);
+        }
+        let results = future::join_all(tasks).await;
+        let mut node = self.node.lock().await;
+        for result in results {
+            if let Some(blob) = result? {
```
Comment (christos-h) [inline linera-core/src/local_node.rs:338]: Wait so it does fail if the blob cannot be downloaded.

File: linera-core/src/node.rs [hunk 3]
```diff
@@ -67,6 +61,28 @@ pub trait ValidatorNode {
     async fn subscribe(&mut self, chains: Vec<ChainId>) -> Result<NotificationStream, NodeError>;
 }
 
+/// Turn an address into a validator node (local node or client to a remote node).
```
Comment (jvff) [inline linera-core/src/node.rs:64]: Nit: maybe remove the "local node" reference?

```suggestion
/// Turn an address into a validator node.
```

**Review by christos-h — APPROVED** [review]

**Review by jvff — APPROVED** [review]

File: linera-core/src/node.rs (hunk 3 above)
Comment (ma2bd) [reply linera-core/src/node.rs:64]: Good catch!

File: linera-core/src/local_node.rs [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+                return info;
+            }
+            let mut result = self.handle_certificate(certificate.clone(), vec![]).await;
+            if let Err(LocalNodeError::WorkerError(WorkerError::ApplicationBytecodesNotFound(
+                locations,
+            ))) = &result
+            {
+                let chain_id = certificate.value().chain_id();
+                let mut blobs = Vec::new();
+                for maybe_blob in future::join_all(locations.iter().map(|location| {
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-core/src/local_node.rs:208]: sure

File: linera-core/src/local_node.rs [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+    where
+        A: ValidatorNode + Send + Sync + 'static + Clone,
+    {
+        let mut info = None;
+        for certificate in certificates {
+            let hash = certificate.hash();
+            if !certificate.value().is_confirmed() || certificate.value().chain_id() != chain_id {
+                // The certificate is not as expected. Give up.
+                tracing::warn!("Failed to process network certificate {}", hash);
+                return info;
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-core/src/local_node.rs:199]: Not sure. I'm just moving this code around

File: linera-core/src/local_node.rs [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+                info.next_block_height,
+                target_next_block_height,
+            )
+            .await?;
+        }
+        let info = self.local_chain_info(chain_id).await?;
+        if target_next_block_height <= info.next_block_height {
+            Ok(info)
+        } else {
+            Err(LocalNodeError::CannotDownloadCertificates {
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-core/src/local_node.rs:299]: We don't have one really, the validators collectively fail to make the expected certs available.

File: linera-core/src/local_node.rs (hunk 1 above)
Comment (ma2bd) [reply linera-core/src/local_node.rs:309]: I leave this for @afck to optimize

File: linera-core/src/local_node.rs (hunk 2 above)
Comment (ma2bd) [reply linera-core/src/local_node.rs:338]: @afck

File: linera-core/src/local_node.rs [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+        mut client: A,
+        chain_id: ChainId,
+        start: BlockHeight,
+        stop: BlockHeight,
+    ) -> Result<(), LocalNodeError>
+    where
+        A: ValidatorNode + Send + Sync + 'static + Clone,
+    {
+        let limit = u64::from(stop)
+            .checked_sub(u64::from(start))
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-core/src/local_node.rs:399]: One could argue the other way.

File: linera-core/src/local_node.rs (hunk 2 above)
Comment (afck) [reply linera-core/src/local_node.rs:338]: `read_or_download_blob` only returns an error if there was a _storage_ error. If the blob couldn't be downloaded it returns `Ok(None)`.

File: linera-core/src/local_node.rs (hunk 1 above)
Comment (afck) [reply linera-core/src/local_node.rs:309]: Good point! I'm fine with either approach.

**Pull request merged**